internal/handlers/watch_cache.go

```diff
@@ -0,0 +1,175 @@
+		fetchFn:            fetchFn,
+		fetchFromTailCache: false,
+	}
+}
+
+func (sub *watchWorker) Close() {
+	if sub.cache == nil {
+		return
+	}
+	sub.cache.manager.mu.Lock()
+	defer sub.cache.manager.mu.Unlock()
+	sub.cache.openCounter -= 1
+	sub.cache = nil
+}
+
+func (sub *watchWorker) Fetch(db *gorm.DB, gtRevision uint64) (WatchableList, error) {
+	if !sub.fetchFromTailCache {
+
+		wl, err := sub.fetchFn(db, gtRevision)
+		if err != nil {
+			return wl, err
+		}
+
+		// empty fetch means we have hit the tail... next fetch will load from the shared tail cache
+		if wl.Len() == 0 {
+			sub.fetchFromTailCache = true
+		}
+
+		return wl, nil
+
+	} else {
+		return sub.fetchFromSharedTailCache(db, gtRevision)
+	}
+}
+
+func (sub *watchWorker) fetchFromSharedTailCache(db *gorm.DB, gtRevision uint64) (WatchableList, error) {
+	for {
+		items, writePos := sub.fetchFromRing(gtRevision)
+		if len(items) != 0 {
+			return items, nil
+		}
+
+		if !sub.fetchFromTailCache {
+			// we can't read from ring buffer anymore, go back to using the Fetch method.
+			return sub.Fetch(db, gtRevision)
+		}
+
+		result, err := sub.tryFillRing(db, gtRevision, writePos)
+		if result != nil || err != nil {
+			return result, err
+		}
+		// we were not the sub that did the DB work, loop around
+		// to see if we can use the results of sub that did do the work.
+	}
+}
+
+func (sub *watchWorker) fetchFromRing(lastRevision uint64) (watchItemList, uint64) {
+	items := watchItemList{}
+
+	// get items of the current cache...
+	cache := sub.cache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	writePos := cache.writePos
+	ringSize := uint64(len(cache.ringBuffer))
+	for sub.readPos != writePos {
+		i := sub.readPos % ringSize
+		item := cache.ringBuffer[i]
+		sub.readPos += 1
+		if lastRevision < item.Revision {
+			if lastRevision+1 == item.Revision {
+				items = append(items, item)
+				lastRevision = item.Revision
+			} else {
+				//fmt.Println("sub.fetchFromTailCache = false")
+				sub.fetchFromTailCache = false
+				break
+			}
+		}
+	}
+	return items, writePos
+}
+
+func (sub *watchWorker) tryFillRing(db *gorm.DB, gtRevision uint64, expectedWritePos uint64) (WatchableList, error) {
+	// multiple subs will compete to fill the cache... only the first in will do the work
+	cache := sub.cache
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	if expectedWritePos == cache.writePos {
+		// fetch the data, and write it to the ring buffer...
+		w, err := sub.fetchFn(db, gtRevision)
+		if err != nil {
+			return w, err
+		}
+		ringSize := uint64(len(cache.ringBuffer))
+
+		fetchLength := w.Len()
+		if fetchLength > int(ringSize) {
+			// we fetched more than what fits in the ring, NOT ideal.
+			fetchLength = int(ringSize)
+		}
+
+		for i := 0; i < fetchLength; i++ {
+			item, revision, deletedAt := w.Item(i)
+			cache.ringBuffer[cache.writePos%ringSize] = watchItem{
+				Item:      item,
+				Revision:  revision,
+				DeletedAt: deletedAt,
+			}
+			cache.writePos += 1
+		}
```

Comment:
It looks like this caching approach can result in multiple copies of the same device sitting in the cache, so an API watcher could get a response with more than one revision of a device. Is that right? Or is something preventing that?

Reply:
We could potentially optimize this. While it wouldn't disrupt the event stream generated by a worker, we could consider implementing deduplication on a per-worker basis to minimize the volume of events transmitted to the API client.

To gauge the practicality of this optimization, it might be beneficial to employ metrics and profiling to assess how frequently duplicates occur in real-world scenarios. This would help us evaluate whether the added CPU overhead for deduplicating device updates justifies the benefits.

Reply:
I'm less concerned about the performance implication and more about the correctness of it. Should an API consumer always be prepared to potentially receive multiple updates for the same device in a single stream of events before a bookmark? I just read back over the code in nexd, and it seems fine, though.

Reply:
OK, while it seems like it doesn't break anything, I still find it really odd that we may send a device multiple historical copies of data. We may send data that was changed hours, days, or weeks ago.

What is the saying? The two hard things in computer science are naming things, cache invalidation, and off by 1 errors? 

I think I'd feel better with a cache invalidation strategy.

Reply:
Yes.  The event consumer has to, and does deal with this dup event case.  You can think of the dup as an invalidation of the previous event.

Reply:
BTW.. even without this patch, dup events could be sent to the consumer since a device could be updated between DB fetches before we get to the tail/bookmark.

Reply:
ah ok, makes sense